redirector: a Go implementation for the root redirector

strib opened the pull request:
This implementation of the root redirector will likely be used for macOS, at least until we figure out the kbfuse+Rust situation.

Note that I needed some new bazil.org/fuse options, so I forked it to github.comkeybase/fuse and added two new features:
* A NoBrowse() option that keeps the redirector from showing up in the Finder (instead, the specific user's mount will show up).
* Adding the requesting user's UID to the context for every request, since `ReadDirAll` needs to know whether the requesting user has a valid mountpoint or not.  The bazil.org/fuse interface for `ReadDirAll` doesn't pass this down, so putting it in the context seemed like the easiest option for now.

Issue: KBFS-2774

oconnor663 commented on redirector/go/main.go:
```diff
@@ -0,0 +1,267 @@
+	}
+	if currUser.Uid != "0" {
+		runtime.LockOSThread()
+		_, _, errNo := syscall.Syscall(syscall.SYS_SETUID, 0, 0, 0)
```
Help me understand what happens if we don't make this call? My Rust redirector doesn't do any SUID step, but if I `chmod 4755` the binary, then I'm able to point it at a regular `0x755 root:root` directory and mount with no problems.

strib replied:
@oconnor663: without this call, if you do `./redirector` as a regular user (without using `sudo`), you get an error.  I.e.:

```sh
$ go build
$ cp go /tmp/redirector
$ cd /tmp/
$ mkdir keybase
$ sudo chown root:root keybase
$ sudo chmod 755 keybase
$ sudo chown root:root redirector
$ sudo chmod 4755 redirector
$ ./redirector /tmp/keybase 
2018/02/28 13:15:29 mount helper error: fusermount: user has no write access to mountpoint /tmp/keybase
Mount error, exiting cleanly: fusermount: exit status 1
```

oconnor663 replied:
Oof, why does this work with a Rust binary but not a Go binary? I just rechecked that it does. (I wasn't able to use `/tmp` for everything though, because it's a `nosuid` mount on my system. Hopefully that doesn't matter here.)

oconnor663 replied:
Another more trivial Go program seems to work just fine with the suid flag. (I just had it create a file in the current directory. Works in root-owned directories without `sudo`.)

strib replied:
Say what?  Yeah you're right, I just tried your branch and it worked for me (in `/tmp/`).  Stumped.

strib replied:
I guess maybe it has something to do with the difference between linking into the .so file and calling an external mount program?  Maybe the latter doesn't get run as root unless you officially set the uid?

oconnor663 replied:
Just tried it with a toy example, and you're right about what happens to child processes with and without setuid. So I believe this theory. Interesting.

oconnor663 replied:
More detail: It looks like a setuid binary does _not_ set the actual `uid`, but rather sets the "effective" uid (that is, the value of `geteuid()`). The latter is used to check file access permissions, but it's not inherited by children.

oconnor663 replied:
(This is filthy and Unix should be ashamed :-D )

oconnor663 replied:
Ok what if instead of calling `syscall.Syscall` (which as I understand it is the main thing preventing us from running this implementation on Linux), what if we used some CGO like this?

```go
/*
#include <sys/types.h>
#include <unistd.h>
#include <errno.h>

int setuid_root() {
	int ret = setuid(0);
	if (ret == 0) {
		return 0;
	} else {
		return errno;
	}
}
*/
import "C"

func setUidRoot() error {
	ret := C.setuid_root()
	if ret != 0 {
		return fmt.Errorf("setuid(0) returned non-zero ERRNO: %d", ret)
	}
	return nil
}
```

If that worked, maybe we'd be fine running the Go redirector on both Linux and Mac. I don't think the security situation is really much different between the two languages.

strib replied:
For the record, we tried the cgo solution for `keybase-mount-helper` and it apparently didn't work: https://github.com/keybase/client/issues/10487#issuecomment-369709787

jzila commented on redirector/go/main.go:
```diff
@@ -0,0 +1,266 @@
+		return "", fuse.ENOENT
+	}
+	if len(fuseMountPoints) == 1 {
+		return fuseMountPoints[0], nil
```
I'm slightly worried about this short circuit: if we somehow have the redirector running but not KBFS, and we have exactly one FUSE mount that isn't KBFS, we'll return it here.

IMO it'd be better to go through the loop below even in the `len(fuseMountPoints) == 1` case.

jzila commented on redirector/go/main.go:
```diff
@@ -0,0 +1,266 @@
+	}
+
+	// Give up and return the first one.
+	return fuseMountPoints[0], nil
```
Shouldn't we return `"", fuse.ENOENT`? We've established at this point that none of the FUSE mounts are KBFS.

strib replied:
My thought on this, and the one above, is that we're not actually sure it's not KBFS, we just know it wasn't mounted with "keybase" in the name.  I was trying not to make that a requirement.  But you're right, it was kinda half-assed.  We can make the naming an official requirement.

strib replied:
Ah, I realize I left the `.kbfs_error` check in this version.  That is wrong; the root user doesn't actually have access to check that in someone else's mount.  So we are forced to go completely from the name.

jzila replied:
Really? The root user doesn't have universal access?

jzila replied:
Is this a mount namespace thing?

strib replied:
You need to give `allow_root` as a mount option, which we don't do, so root can't, say, index your files.

jzila replied:
We could make a redirector for the redirector, which redirects to a mount that root has access to but maps 1:1 to user mounts. Redirects all the way down!

</joking>

jzila commented on redirector/go/main.go:
```diff
@@ -0,0 +1,266 @@
+		"kbfs.error.txt", "kbfs.nologin.txt", ".kbfs_enable_auto_journals",
+		".kbfs_disable_auto_journals", ".kbfs_enable_block_prefetching",
+		".kbfs_disable_block_prefetching", ".kbfs_enable_debug_server",
+		".kbfs_disable_debug_server":
```
Might be nice to export this set in one of our libraries.

strib replied:
I didn't want to suck `libfs` and `libkbfs` into this binary, and I didn't feel like making a new one-file library for this.  But if you feel strongly about it, let me know.

jzila commented on redirector/go/main.go:
```diff
@@ -0,0 +1,266 @@
+	}
+
+	interruptChan := make(chan os.Signal, 1)
+	signal.Notify(interruptChan, os.Interrupt)
```
Can we do this on a `SIGTERM` as well?

songgao replied:
SIGTERM is actually probably more often since that's the default for `kill`/`killall`

oconnor663 commented:
Added a comment above about CGO.

songgao commented on redirector/go/main.go:
```diff
@@ -0,0 +1,266 @@
+			continue
+		}
+		// Double-check that this is a real KBFS mount.
+		if _, err := os.Stat(filepath.Join(mp, ".kbfs_error")); err != nil {
```
Curious why not `.kbfs_status`? Is this one cheaper?

strib replied:
Yeah much cheaper, but anyway, I removed this.

strib commented on redirector/go/main.go:
```diff
@@ -0,0 +1,266 @@
+		return ""
+	}
+	now := time.Now()
+	if now.Sub(entry.time) > mountpointTimeout {
```
In Go, `time.Sub()` uses the monotonic clock, so wallclock and timezone changes aren't relevant.  From https://golang.org/pkg/time/#hdr-Monotonic_Clocks:
> If Times t and u both contain monotonic clock readings, the operations t.After(u), t.Before(u), t.Equal(u), and t.Sub(u) are carried out using the monotonic clock readings alone, ignoring the wall clock readings.

jzila replied:
Oh right I forgot they made that change!

akalin-keybase commented on vendor/github.com/keybase/gomounts/darwin.go:
```diff
@@ -0,0 +1,47 @@
+	result := make([]Volume, 0)
+
+	var mntbuf *C.struct_statfs
+	count := int(C.getmntinfo(&mntbuf, C.MNT_NOWAIT))
```
this code is buggy:

- `getmntinfo` return *0* on an error and sets `errno`.
- would feel much better using the thread-safe version `getmntinfo_r_np` that doesn't shrae a buffer.
- the code casting the array to a slice below is kinda scary -- https://golang.org/pkg/reflect/#SliceHeader says `Data` might still be garbage collected? would much rather just iterate manually somehow.

akalin-keybase commented on vendor/github.com/keybase/gomounts/linux.go:
```diff
@@ -0,0 +1,45 @@
+		uidstr := ""
+		if substr := C.GoString(C.hasmntopt(ent, mntopt)); len(substr) > 0 {
+			commasplit := strings.SplitN(substr, ",", 2)
+			equalsplit := strings.SplitN(commasplit[0], "=", 2)
```
this seems a bit brittle to me, i'd have to test to see if it handles edge cases ok

akalin-keybase replied:
alternatively, we could just look at what `setmntopt` does and mimick its behavior.

strib replied:
@akalin-keybase what is `setmntopt`?  Google shows nothing.

I think this code is pretty straightforward.  `hasmntopt` returns a substring starting at "user_id=", possibly with comma-separated options after it.  So we want the string between the first equal sign and the first comma.

Unless you have specific concerns, I'm not going to spend any more time on this part.

akalin-keybase replied:
sorry, meant `setmntent`

strib replied:
I don't think `setmntent` does anything with options, it just opens a file: https://github.com/seL4/libmuslc/blob/master/src/linux/mntent.c#L8-L10

I'm going to leave this as is for now until we discover an issue with it.

akalin-keybase commented on vendor/github.com/keybase/gomounts/linux.go:
```diff
@@ -0,0 +1,45 @@
+	var ent *C.struct_mntent
+
+	mntopt := C.CString("user_id")
+	for ent = C.getmntent(file); ent != nil; ent = C.getmntent(file) {
```
would prefer `getmntent_r` here, too

strib replied:
I've been struggling with converting this code to `getmntent_r` for a couple hours now, and have failed.  Basically it is very not obvious how to pass a `struct mntent *` to an stack-allocated `C.struct_mntent` into C-land, without rewriting the entire function is C.

Instead, I'm going to give up and just protect all calls to `GetMountedVolumes()` with a global lock in `main.go`.

akalin-keybase commented on vendor/github.com/keybase/gomounts/windows.go:
```diff
@@ -0,0 +1,40 @@
+// Windows implementation
+func getMountedVolumes() ([]Volume, error) {
+	result := make([]Volume, 0)
+	var buf [256]C.char
```
should be `MAX_PATH + 1`, which is 261

akalin-keybase commented on vendor/github.com/keybase/gomounts/windows.go:
```diff
@@ -0,0 +1,40 @@
+	result := make([]Volume, 0)
+	var buf [256]C.char
+
+	drives := uint32(C.GetLogicalDrives())
```
hmm, isn't it possible for mounts to be places other than `<LETTER>:\`? @zanderz do you know?

akalin-keybase replied:
Not that it matters too much, since we won't be running this on windows

strib commented on redirector/go/main.go:
```diff
@@ -0,0 +1,266 @@
+
+func (r *root) findKBFSMount(ctx context.Context) (
+	mountpoint string, err error) {
+	uid := ctx.Value(fs.CtxHeaderUIDKey).(uint32)
```
I actually wanted a crash here if it's not set.  It means it's compiled with a wrong version of bazil and we'd have big problems. I'll add a comment.

songgao replied:
Got it. Comment is good enough for me!

strib commented on redirector/go/main.go:
```diff
@@ -0,0 +1,266 @@
+	mountpoint string, err error) {
+	uid := ctx.Value(fs.CtxHeaderUIDKey).(uint32)
+	// Don't let the root see anything here; we don't want a symlink
+	// loop back to this mount.
```
> Not sure if making a root-started KBFS work with redirector is a goal though.

Not a goal.

strib replied:
(The keybase service refuses to even run as root.)

jzila commented on redirector/go/main.go:
```diff
@@ -56,14 +56,35 @@ type cacheEntry struct {
+	// Find mountpoints like "/home/user/.local/share/keybase/fs", or
+	// "/home/user/keybase", and make sure it doesn't match mounts for
+	// another run mode, say "/home/user/keybase.staging".
+	rx, err := regexp.Compile(fmt.Sprintf("/%s/|%s$", runmodeStr))
```
This `Sprintf` takes 2 strings, but only passes in one.

jzila replied:
Additionally, the runmode string can have a `.` in it, but in regex land that means any character.

strib commented:
> I'm not a fan of using a regex in a setuid root application. Regexes are notorious sources of security holes and bugs. We should use them sparingly. Can you use strings.IndexOf instead?

Ok done.  I think the Go regexp is pretty safe and the new code is harder to understand though.

jzila commented:
Thanks! I agree that the check is a bit harder to read but I still think it's lower risk.

strib commented:
Ok I think I've addressed all the comments I'm qualified to address.  I'll merge when CI passes, but if there are follow-on comments I'll fix them in a separate PR.  Thanks all!
